### PR TITLE
Refactor user group to new programming model

### PR DESCRIPTION
**Pull request by stonezdj**
Add context to required methods
Add pkg/usergroup/dao
Add pkg/usergroup/manager
Add controller/usergroup/controller

Signed-off-by: stonezdj <stonezdj@gmail.com>

### CODE REVIEW

File: api/v2.0/swagger.yaml [hunk 1]
```diff
@@ -2230,6 +2230,131 @@ paths:
           $ref: '#/responses/403'
         '500':
           $ref: '#/responses/500'
+  /usergroups:
+    get:
+      summary: Get all user groups information
+      description: Get all user groups information
+      tags:
+        - usergroup
```
Comment (heww) [inline api/v2.0/swagger.yaml:2238]: use `operationId: listUserGoups` to this API

File: api/v2.0/swagger.yaml [hunk 2]
```diff
@@ -2230,6 +2230,131 @@ paths:
...
+          $ref: '#/responses/401'
+        '403':
+          $ref: '#/responses/403'
+        '500':
+          $ref: '#/responses/500'
+    post:
+      summary: Create user group
+      description: Create user group information
+      tags:
+        - usergroup
```
Comment (heww) [inline api/v2.0/swagger.yaml:2256]: use `operationId: createUserGroup` to this API

File: api/v2.0/swagger.yaml [hunk 3]
```diff
@@ -2230,6 +2230,131 @@ paths:
...
+        '409':
+          $ref: '#/responses/409'
+        '500':
+          $ref: '#/responses/500'
+  '/usergroups/{group_id}':
+    get:
+      summary: Get user group information
+      description: Get user group information
+      tags:
+        - usergroup
```
Comment (heww) [inline api/v2.0/swagger.yaml:2284]: `operationId: getUserGroup` ?

File: api/v2.0/swagger.yaml [hunk 4]
```diff
@@ -2230,6 +2230,131 @@ paths:
...
+          $ref: '#/responses/403'
+        '404':
+          $ref: '#/responses/404'
+        '500':
+          $ref: '#/responses/500'
+    put:
+      summary: Update group information
+      description: Update user group information
+      tags:
+        - usergroup
```
Comment (heww) [inline api/v2.0/swagger.yaml:2311]: `operationId: updateUserGroup` ?

File: api/v2.0/swagger.yaml [hunk 5]
```diff
@@ -2230,6 +2230,131 @@ paths:
...
+          $ref: '#/responses/403'
+        '404':
+          $ref: '#/responses/404'
+        '500':
+          $ref: '#/responses/500'
+    delete:
+      summary: Delete user group
+      description: Delete user group
+      tags:
+        - usergroup
```
Comment (heww) [inline api/v2.0/swagger.yaml:2341]: `operationId: deleteUserGroup` ?

File: api/v2.0/swagger.yaml (hunk 3 above)
Comment (heww) [reply api/v2.0/swagger.yaml:2284]: include the `requestId` in the parameters?

File: api/v2.0/swagger.yaml (hunk 4 above)
Comment (heww) [reply api/v2.0/swagger.yaml:2311]: include the `requestId` in the parameters?

File: api/v2.0/swagger.yaml (hunk 5 above)
Comment (heww) [reply api/v2.0/swagger.yaml:2341]: include the `requestId` in the parameters?

File: api/v2.0/swagger.yaml (hunk 2 above)
Comment (heww) [reply api/v2.0/swagger.yaml:2256]: include the `requestId` in the parameters?

File: api/v2.0/swagger.yaml (hunk 1 above)
Comment (heww) [reply api/v2.0/swagger.yaml:2238]: include the `requestId` in the parameters?

File: api/v2.0/swagger.yaml (hunk 4 above)
Comment (stonezdj) [reply api/v2.0/swagger.yaml]: Updated

File: api/v2.0/swagger.yaml (hunk 5 above)
Comment (stonezdj) [reply api/v2.0/swagger.yaml]: Updated

File: api/v2.0/swagger.yaml (hunk 3 above)
Comment (stonezdj) [reply api/v2.0/swagger.yaml]: Updated

File: api/v2.0/swagger.yaml (hunk 2 above)
Comment (stonezdj) [reply api/v2.0/swagger.yaml]: Updated

File: api/v2.0/swagger.yaml (hunk 1 above)
Comment (stonezdj) [reply api/v2.0/swagger.yaml]: Updated

File: src/server/v2.0/handler/usergroup.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+	ug := model.UserGroup{
+		GroupName:   params.Usergroup.GroupName,
+		GroupType:   int(params.Usergroup.GroupType),
+		LdapGroupDN: params.Usergroup.LdapGroupDn,
+	}
+	id, err := u.ctl.Create(ctx, ug)
+	if err != nil {
+		return u.SendError(ctx, err)
+	}
+	return operation.NewCreateUserGroupCreated().WithLocation(fmt.Sprintf("/api/v2.0/usergroups/%d", id))
```
Comment (heww) [inline src/server/v2.0/handler/usergroup.go:59]: `location = fmt.Sprintf("%s/%d", strings.TrimSuffix(params.HTTPRequest.URL.Path, "/"), id)`

File: src/server/v2.0/handler/usergroup.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+		LdapGroupDn: ug.LdapGroupDN,
+	}
+	return operation.NewGetUserGroupOK().WithPayload(userGroup)
+}
+
+func (u userGroupAPI) ListUserGroups(ctx context.Context, params operation.ListUserGroupsParams) middleware.Responder {
+	if err := u.RequireSystemAccess(ctx, rbac.ActionList, rbac.ResourceUserGroup); err != nil {
+		return u.SendError(ctx, err)
+	}
+	// TODO: replace with the new configure api
```
Comment (heww) [inline src/server/v2.0/handler/usergroup.go:102]: new configure API was merged

**Review by heww — APPROVED** [review]

File: src/controller/usergroup/controller.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+func (c *controller) Populate(ctx context.Context, userGroups []model.UserGroup) ([]int, error) {
+	return c.mgr.Populate(ctx, userGroups)
+}
+
+func (c *controller) Ensure(ctx context.Context, group *model.UserGroup) error {
+	return c.mgr.Onboard(ctx, group)
+}
+
+func newController() Controller {
```
Comment (wy65701436) [inline src/controller/usergroup/controller.go:66]: pop this method

File: src/server/v2.0/handler/usergroup.go (hunk 7 above)
Comment (stonezdj) [reply src/server/v2.0/handler/usergroup.go]: Updated

**Review by wy65701436 — APPROVED** [review]
lgtm

File: src/controller/usergroup/controller.go (hunk 8 above)
Comment (stonezdj) [reply src/controller/usergroup/controller.go:66]: updated

**Pull request merged**